Synced node gets unsynced and gives error: building past consensus base mana vector: %vbase mana should never be negative

## Bug description

I created a new node which got synced fairly quick... but not long there after got unsynced with this error:
`ERROR Mana messagelayer/mana_plugin.go:686 error building past consensus base mana vector: %vbase mana should never be negative`

Btw: the database folder contained 20 items and was 1.75GB in size which seems high(?) for just one day uptime.

## GoShimmer version

v0.5.6

## Hardware specification

Docker on Synology NAS DSM 6

## Steps To reproduce the bug

In my case I just created a new node after deleting the db folder.
It's reproducable. Tried it three times. Same error. See second log which includes the first.
By the way the third time my node did stay in sync for ~6 hours while giving the mana errors before finally getting unsynced.

## Expected behaviour

A synced node without any errors in log

## Actual behaviour

Most of the time an unsynced node

## Errors

`ERROR Mana messagelayer/mana_plugin.go:686 error building past consensus base mana vector: %vbase mana should never be negative`

Full log attached.

[goshimmer-1.html.txt](https://github.com/iotaledger/goshimmer/files/6276453/goshimmer-1.html.txt)

[goshimmer-2.html.txt](https://github.com/iotaledger/goshimmer/files/6277244/goshimmer-2.html.txt)

--- a/packages/mana/consensusbase.go
+++ b/packages/mana/consensusbase.go
@@ -28,14 +28,13 @@ func (c *ConsensusBaseMana) updateEBM1(n time.Duration) {
 	if c.BaseMana1 == c.EffectiveBaseMana1 {
 		return
 	}
-	// they are not the same, but close. Stop the future updates.
-	if math.Abs(c.BaseMana1-c.EffectiveBaseMana1) < DeltaStopUpdate {
-		c.EffectiveBaseMana1 = c.BaseMana1
-		return
-	}
 	// normal update
 	c.EffectiveBaseMana1 = math.Pow(math.E, -emaCoeff1*n.Seconds())*c.EffectiveBaseMana1 +
 		(1-math.Pow(math.E, -emaCoeff1*n.Seconds()))*c.BaseMana1
+	// they are not the same, but close. Stop the future updates.
+	if math.Abs(c.BaseMana1-c.EffectiveBaseMana1) < DeltaStopUpdate {
+		c.EffectiveBaseMana1 = c.BaseMana1
+	}
 }
 
 func (c *ConsensusBaseMana) revoke(amount float64, t time.Time) error {

--- a/packages/mana/events.go
+++ b/packages/mana/events.go
@@ -7,6 +7,8 @@ import (
 
 	"github.com/iotaledger/hive.go/events"
 	"github.com/iotaledger/hive.go/identity"
+	"github.com/iotaledger/hive.go/stringify"
+	"github.com/mr-tron/base58"
 
 	"github.com/iotaledger/goshimmer/packages/ledgerstate"
 )
@@ -53,14 +55,16 @@ const (
 
 // Event is the interface definition of an event.
 type Event interface {
-	// ManaType returns the type of the event.
+	// Type returns the type of the event.
 	Type() byte
 	// ToJSONSerializable returns a struct that can be serialized into JSON object.
 	ToJSONSerializable() interface{}
 	// ToPersistable returns an event that can be persisted.
 	ToPersistable() *PersistableEvent
-	// Time returns the time of the event.
+	// Timestamp returns the time of the event.
 	Timestamp() time.Time
+	// String returns a human readable version of the event.
+	String() string
 }
 
 // PledgedEvent is the struct that is passed along with triggering a Pledged event.
@@ -90,6 +94,18 @@ func (p *PledgedEvent) ToJSONSerializable() interface{} {
 		TxID:     p.TransactionID.Base58(),
 		Amount:   p.Amount,
 	}
+}
+
+// String returns a human readable version of the event.
+func (p *PledgedEvent) String() string {
+	return stringify.Struct("PledgeEvent",
+		stringify.StructField("type", p.ManaType.String()),
+		stringify.StructField("shortNodeID", p.NodeID.String()),
+		stringify.StructField("fullNodeID", base58.Encode(p.NodeID.Bytes())),
+		stringify.StructField("time", p.Time.String()),
+		stringify.StructField("amount", p.Amount),
+		stringify.StructField("txID", p.TransactionID),
+	)
 }
 
 // ToPersistable returns an event that can be persisted.
@@ -174,6 +190,19 @@ func (r *RevokedEvent) ToJSONSerializable() interface{} {
 	}
 }
 
+// String returns a human readable version of the event.
+func (r *RevokedEvent) String() string {
+	return stringify.Struct("RevokedEvent",
+		stringify.StructField("type", r.ManaType.String()),
+		stringify.StructField("shortNodeID", r.NodeID.String()),
+		stringify.StructField("fullNodeID", base58.Encode(r.NodeID.Bytes())),
+		stringify.StructField("time", r.Time.String()),
+		stringify.StructField("amount", r.Amount),
+		stringify.StructField("txID", r.TransactionID),
+		stringify.StructField("inputID", r.InputID),
+	)
+}
+
 // ToPersistable returns an event that can be persisted.
 func (r *RevokedEvent) ToPersistable() *PersistableEvent {
 	return &PersistableEvent{
@@ -238,6 +267,17 @@ func (u *UpdatedEvent) ToJSONSerializable() interface{} {
 			LastUpdated:       u.NewMana.LastUpdate().Unix(),
 		},
 	}
+}
+
+// String returns a human readable version of the event.
+func (u *UpdatedEvent) String() string {
+	return stringify.Struct("UpdatedEvent",
+		stringify.StructField("type", u.ManaType.String()),
+		stringify.StructField("shortNodeID", u.NodeID.String()),
+		stringify.StructField("fullNodeID", base58.Encode(u.NodeID.Bytes())),
+		stringify.StructField("oldBaseMana", u.OldMana),
+		stringify.StructField("newBaseMana", u.NewMana),
+	)
 }
 
 // ToPersistable converts the event to a persistable event.

--- a/plugins/messagelayer/mana_plugin.go
+++ b/plugins/messagelayer/mana_plugin.go
@@ -688,7 +688,7 @@ func pruneConsensusEventLogsStorage() {
 	}
 	manaLogger.Infof("going to prune %d events", i)
 	toBePrunedEvents := make(mana.EventSlice, i)
-	for k := 0; k < i+1; k++ {
+	for k := 0; k < i; k++ {
 		toBePrunedEvents[k] = eventLogs[k]
 	}
 	t := toBePrunedEvents[len(toBePrunedEvents)-1].Timestamp()

--- a/plugins/messagelayer/mana_plugin.go
+++ b/plugins/messagelayer/mana_plugin.go
@@ -681,15 +681,19 @@ func pruneConsensusEventLogsStorage() {
 	prev := eventLogs[slidingEventsInterval-1]
 	var i int
 	for i = slidingEventsInterval; i < len(eventLogs); i++ {
-		if eventLogs[i].Timestamp() != prev.Timestamp() {
+		if !eventLogs[i].Timestamp().Equal(prev.Timestamp()) {
 			break
 		}
 		prev = eventLogs[i]
 	}
-	eventLogs = eventLogs[:i]
-	t := eventLogs[len(eventLogs)-1].Timestamp()
+	manaLogger.Infof("going to prune %d events", i)
+	toBePrunedEvents := make(mana.EventSlice, i)
+	for k := 0; k < i+1; k++ {
+		toBePrunedEvents[k] = eventLogs[k]
+	}
+	t := toBePrunedEvents[len(toBePrunedEvents)-1].Timestamp()
 
-	err = cbmvPast.BuildPastBaseVector(eventLogs, t)
+	err = cbmvPast.BuildPastBaseVector(toBePrunedEvents, t)
 	if err != nil {
 		manaLogger.Errorf("error building past consensus base mana vector: %w", err)
 		return
@@ -716,11 +720,13 @@ func pruneConsensusEventLogsStorage() {
 	consensusBaseManaPastVectorMetadataStorage.Store(metadata).Release()
 
 	var entriesToDelete [][]byte
-	for _, ev := range eventLogs {
+	for _, ev := range toBePrunedEvents {
 		entriesToDelete = append(entriesToDelete, ev.ToPersistable().ObjectStorageKey())
 	}
+	manaLogger.Infof("deleting %d events from consensus event storage", len(entriesToDelete))
 	consensusEventsLogStorage.DeleteEntriesFromStore(entriesToDelete)
 	consensusEventsLogsStorageSize.Sub(uint32(len(entriesToDelete)))
+	manaLogger.Infof("there are %d events remaining in consensus event storage", consensusEventsLogsStorageSize.Load())
 }
 
 func cleanupManaVectors() {

--- a/plugins/messagelayer/mana_plugin.go
+++ b/plugins/messagelayer/mana_plugin.go
@@ -31,8 +31,8 @@ const (
 	// PluginName is the name of the mana plugin.
 	PluginName = "Mana"
 
-	maxConsensusEventsInStorage = 108000
-	slidingEventsInterval       = 10800 // 10% of maxConsensusEventsInStorage
+	maxConsensusEventsInStorage = 110000
+	slidingEventsInterval       = 10000 // 10% of maxConsensusEventsInStorage
 )
 
 var (
@@ -90,6 +90,7 @@ func configureManaPlugin(*node.Plugin) {
 	}
 	consensusEventsLogStorage = osFactory.New(mana.PrefixEventStorage, mana.FromEventObjectStorage)
 	consensusEventsLogsStorageSize.Store(getConsensusEventLogsStorageSize())
+	manaLogger.Infof("read %d mana events from storage", consensusEventsLogsStorageSize.Load())
 	consensusBaseManaPastVectorStorage = osFactory.New(mana.PrefixConsensusPastVector, mana.FromObjectStorage)
 	consensusBaseManaPastVectorMetadataStorage = osFactory.New(mana.PrefixConsensusPastMetadata, mana.FromMetadataObjectStorage)
 
@@ -621,7 +622,8 @@ func pruneConsensusEventLogsStorage() {
 	if consensusEventsLogsStorageSize.Load() < maxConsensusEventsInStorage {
 		return
 	}
-	cachedObj := consensusBaseManaPastVectorMetadataStorage.Get([]byte(mana.ConsensusBaseManaPastVectorMetadataStorageKey))
+
+	cachedObj := consensusBaseManaPastVectorMetadataStorage.Load([]byte(mana.ConsensusBaseManaPastVectorMetadataStorageKey))
 	cachedMetadata := &mana.CachedConsensusBasePastManaVectorMetadata{CachedObject: cachedObj}
 	defer cachedMetadata.Release()
 
@@ -676,21 +678,28 @@ func pruneConsensusEventLogsStorage() {
 		return
 	}
 	eventLogs.Sort()
-	// Make sure to take related events.
+	// we always want (maxConsensusEventsInStorage - slidingEventsInterval) number of events left
+	deleteWindow := len(eventLogs) - (maxConsensusEventsInStorage - slidingEventsInterval)
+	storageSizeInt := int(consensusEventsLogsStorageSize.Load())
+	if deleteWindow < 0 || deleteWindow > storageSizeInt {
+		manaLogger.Errorf("invalid delete window %d for storage size %d, max storage size %d and sliding interval %d",
+			deleteWindow, storageSizeInt, maxConsensusEventsInStorage, slidingEventsInterval)
+		return
+	}
+	// Make sure to take related events. (we take deleteWindow oldest events)
 	// Ensures that related events (same time) are not split between different intervals.
-	prev := eventLogs[slidingEventsInterval-1]
+	prev := eventLogs[deleteWindow-1]
 	var i int
-	for i = slidingEventsInterval; i < len(eventLogs); i++ {
+	for i = deleteWindow; i < len(eventLogs); i++ {
 		if !eventLogs[i].Timestamp().Equal(prev.Timestamp()) {
 			break
 		}
 		prev = eventLogs[i]
 	}
-	manaLogger.Infof("going to prune %d events", i)
-	toBePrunedEvents := make(mana.EventSlice, i)
-	for k := 0; k < i; k++ {
-		toBePrunedEvents[k] = eventLogs[k]
-	}
+	toBePrunedEvents := eventLogs[:i]
+	// TODO: later, when we have epochs, we have to make sure that `t` is before the epoch to be "finalized" next.
+	// Otherwise, we won't be able to calculate the consensus mana for that epoch because we already pruned the events
+	// leading up to it.
 	t := toBePrunedEvents[len(toBePrunedEvents)-1].Timestamp()
 
 	err = cbmvPast.BuildPastBaseVector(toBePrunedEvents, t)
@@ -726,7 +735,7 @@ func pruneConsensusEventLogsStorage() {
 	manaLogger.Infof("deleting %d events from consensus event storage", len(entriesToDelete))
 	consensusEventsLogStorage.DeleteEntriesFromStore(entriesToDelete)
 	consensusEventsLogsStorageSize.Sub(uint32(len(entriesToDelete)))
-	manaLogger.Infof("there are %d events remaining in consensus event storage", consensusEventsLogsStorageSize.Load())
+	manaLogger.Infof("%d events remaining in consensus event storage", consensusEventsLogsStorageSize.Load())
 }
 
 func cleanupManaVectors() {

--- a/plugins/messagelayer/mana_plugin.go
+++ b/plugins/messagelayer/mana_plugin.go
@@ -654,14 +654,17 @@ func pruneConsensusEventLogsStorage() {
 		cachedPe := &mana.CachedPersistableEvent{CachedObject: cachedObject}
 		defer cachedPe.Release()
 		pe := cachedPe.Unwrap()
-		metadata := cachedMetadata.Unwrap()
-
 		var ev mana.Event
 		ev, err = mana.FromPersistableEvent(pe)
-		if ev.Timestamp().Before(metadata.Timestamp) {
-			manaLogger.Errorf("consensus event storage contains event that is older, than the stored metadata timestamp %s: %s", metadata.Timestamp, ev.String())
-			return true
+
+		if cachedMetadata.Exists() {
+			metadata := cachedMetadata.Unwrap()
+			if ev.Timestamp().Before(metadata.Timestamp) {
+				manaLogger.Errorf("consensus event storage contains event that is older, than the stored metadata timestamp %s: %s", metadata.Timestamp, ev.String())
+				return true
+			}
 		}
+
 		if err != nil {
 			return false
 		}

--- a/plugins/messagelayer/mana_plugin.go
+++ b/plugins/messagelayer/mana_plugin.go
@@ -654,9 +654,14 @@ func pruneConsensusEventLogsStorage() {
 		cachedPe := &mana.CachedPersistableEvent{CachedObject: cachedObject}
 		defer cachedPe.Release()
 		pe := cachedPe.Unwrap()
+		metadata := cachedMetadata.Unwrap()
 
 		var ev mana.Event
 		ev, err = mana.FromPersistableEvent(pe)
+		if ev.Timestamp().Before(metadata.Timestamp) {
+			manaLogger.Errorf("consensus event storage contains event that is older, than the stored metadata timestamp %s: %s", metadata.Timestamp, ev.String())
+			return true
+		}
 		if err != nil {
 			return false
 		}
@@ -683,7 +688,7 @@ func pruneConsensusEventLogsStorage() {
 
 	err = cbmvPast.BuildPastBaseVector(eventLogs, t)
 	if err != nil {
-		manaLogger.Error("error building past consensus base mana vector: %v", err)
+		manaLogger.Errorf("error building past consensus base mana vector: %w", err)
 		return
 	}
 
@@ -701,12 +706,11 @@ func pruneConsensusEventLogsStorage() {
 		Timestamp: t,
 	}
 
-	if !cachedMetadata.Exists() {
-		consensusBaseManaPastVectorMetadataStorage.Store(metadata).Release()
-	} else {
-		m := cachedMetadata.Unwrap()
-		m.Update(metadata)
+	if err = consensusBaseManaPastVectorMetadataStorage.Prune(); err != nil {
+		manaLogger.Errorf("error pruning consensus base mana vector metadata storage: %w", err)
+		return
 	}
+	consensusBaseManaPastVectorMetadataStorage.Store(metadata).Release()
 
 	var entriesToDelete [][]byte
 	for _, ev := range eventLogs {

